### ISSUE TITLE
Issue with asp.net postback

### ISSUE THREAD

**Issue by vchopra88** [apereo/dotnet-cas-client#43]
**Steps to Repro**
Create a simple website, add a web page (webpage1.aspx) which can demonstrate a simple postback like retrieving isPostback value. 
Configure the website .net CAS
Publish the website to IIS.
Set the default document to webpage1.aspx.

**Workaround**
As of now implemented a workaround by creating a default.aspx and redirecting to webpage1.aspx.

**Ask**
Is there CAS configuration where we can redirect the user to webpage1.aspx after login or Is there known issue with IIS default document module.

**Comment by randyh360** [comment]
I've been updating a bunch of applications from IIS 6 to IIS 8.5 with integrated mode enabled lately and have run into this problem. You can take a look at the following [stackoverflow question](http://stackoverflow.com/questions/7228344/postback-doesnt-work-with-aspx-page-as-default-document/15353347) for more info, but basically you either have to put "webpage1.aspx" into your form's action attribute or add the following to global.asax:

```
void Application_BeginRequest(object sender, EventArgs e)
{
    var app = (HttpApplication)sender;
    if (app.Context.Request.Url.LocalPath.EndsWith("/"))
    {
    app.Context.RewritePath(
             string.Concat(app.Context.Request.Url.LocalPath, "webpage1.aspx"));
    }
}
```

**Comment by vchopra88** [comment]
Thanks @randyh360 for providing information about modifying 'Application_BeginRequest', 'IsPostBack' value works okay now!

**Comment by TheHokieCoder** [comment]
The core issue here is that the IIS default document module performs a rewrite of the requested URL. This means that when a request path of `/` is encountered, IIS rewrites the request (internally for its routing mechanisms) as `/webpage1.aspx` (using **vchopra88**'s repro). Unfortunately, the rewritten request path value never makes it to the page's form `action` attribute, which ends up breaking the postback mechanism since the postback request will have a different request path from the currently executing page object. There are several ways to fix this, as **gitrandynow** mentioned above, but they all fix the symptom, not the problem.

Bottom line, this is a web server issue, not a DotNetCasClient issue, so I am closing the issue. Additionally, I would think that it would be out of the client's scope to try to provide a way around this since the client is in the business of security, not of routing web requests for an application.

As an aside for future viewers of this issue, I, too, have encountered this issue and solved it inside the code-behind for my master page(s). Inside the `PreRender` event, I set the form's `.Action` property to `Request.url.PathAndQuery` when it is empty or ends with a "/". It works well for me since I use masters exclusively in my applications.

**Comment by phantomtypist** [comment]
Forgive my ignorance, but is this only an issue with WebForms?  I don't believe my group has ever experienced this problem with MVC.

**Comment by TheHokieCoder** [comment]
@phantomtypist I believe it would only rear itself when using WebForms because MVC is 100% based on routes (e.g. `/{controller}/{action}`) whereas WebForms is based (mostly) on documents. The default document module emulates a route by taking the request to `/` and "mapping" it to `/webpage1.aspx`, but in doing so something forgets to update the HTML form server side to the new URL.

It's been a while since I've done MVC, so I don't really remember the specifics of how MVC handles postbacks, viewstate, etc. But it would seem that there is no postback in MVC like there is in WebForms since the controller methods are decorated with HTTP method attributes like GET, POST, etc., right? In other words, I don't think MVC posts back to itself like webforms do...rather it gets redirected to a different controller method based on the route and method.

**Comment by phantomtypist** [comment]
Yeah, MVC doesn't really have any concept like PostBack nor ViewState (at least in the WebForms sense.)

Thanks for the explanation.  I just wasn't sure if this problem manifested itself in a different way on MVC or not.
